### PR TITLE
Drive states

### PR DESCRIPTION
**Pull request by SuuperW**
Add support for putting TAStudio's state history buffers and/or ancients in files instead of keeping it all in RAM.

### CODE REVIEW

**Comment by nattthebear** [comment]
Would it be simpler and more effective to simply allow the backing address space for ZWinderBuffer to be a mmapped file?

File: src/BizHawk.Client.Common/movie/tasproj/ZwinderStateManager.cs [hunk 1]
```diff
@@ -41,24 +49,41 @@ public ZwinderStateManager(Func<int, bool> reserveCallback)
...
 		{
+			SetId();
 			_current = current;
 			_recent = recent;
 			_gapFiller = gapFiller;
 			_ancientInterval = ancientInterval;
 			_reserveCallback = reserveCallback;
 		}
 		
+		private void SetId()
```
Comment (nattthebear) [inline src/BizHawk.Client.Common/movie/tasproj/ZwinderStateManager.cs:76]: This needs to be threadsafe -- the usual expectation in C# is that while single objects aren't threadsafe, multiple threads can independently create independent objects without conflict.  But here the internal implementation detail hits a static resource without synchronization.  Try `Interlocked.Increment`.

File: src/BizHawk.Client.Common/rewind/ZwinderBuffer.cs [hunk 2]
```diff
@@ -18,6 +18,13 @@ public class ZwinderBuffer : IDisposable
 		*/
```
Comment (nattthebear) [inline src/BizHawk.Client.Common/rewind/ZwinderBuffer.cs:18]: The general problem I have in this file is that ZwinderBuffer is made to do one thing: abstract on top of a ring buffer that's mapped to memory.  With the UseDrive stuff, a bunch of important methods have completely unrelated separate implementations for the two cases.  I don't see that this should all be handled in one file.  If there is commonality, it should be brought out to a base class or base interface or something,

**Comment by zeromus** [comment]
I spent a full 2 hours thinking about this and have a number of ideas for how it could be done. But the lowest-impact, safest, and simplest idea: Can we have SaveStateStream and LoadStateStream use an underlying stream indiscriminately after doing the ring buffer logic? Then we can make the underlying stream a filestream or a new slim little stream that simply sits on top of a raw pointer, if there isn't already such a thing. I think I could do this in about 1 more hour.

**Comment by nattthebear** [comment]
> I spent a full 2 hours thinking about this and have a number of ideas for how it could be done. But the lowest-impact, safest, and simplest idea: Can we have SaveStateStream and LoadStateStream use an underlying stream indiscriminately after doing the ring buffer logic? Then we can make the underlying stream a filestream or a new slim little stream that simply sits on top of a raw pointer, if there isn't already such a thing. I think I could do this in about 1 more hour.

I like that.  Take a look at `MemoryBlock.GetStream`, that should do what we need.

**Comment by SuuperW** [comment]
I also like the idea of accessing the memory block as a stream. I had considered making the two share an interface, but for some reason I only thought of trying to make the file accessed more like the memory block, but that didn't seem very clean either. Going the other way and turning the memory block into a stream should be super easy and clean, since we already have MemoryBlock.GetStream.

**Comment by nattthebear** [comment]
I would like to see this refactored to use the capability added in https://github.com/TASVideos/BizHawk/pull/2536.  It's OK if some additional changes are needed to ZwinderBuffer above and beyond what that does, but hopefully most of that is already taken care of.

**Pull request closed without merging**